Std/Data/ListM/Basic.lean

```diff
@@ -0,0 +1,421 @@
+/-
+Copyright (c) 2018 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro, Keeley Hoek, Simon Hudon, Scott Morrison
+-/
+
+/-! # Monadic lazy lists.
+
+Lazy lists with "laziness" controlled by an arbitrary monad.
+-/
+
+
+/-!
+In an initial section we describe the specification of `ListM`,
+and provide a private unsafe implementation,
+and then a public `opaque` wrapper of this implementation, satisfying the specification.
+-/
+namespace ListM
+
+variable (m : Type u → Type u)
+
+private structure Spec where
+  listM : Type u → Type u
+  nil : listM α
+  cons' : α → listM α → listM α
+  squash : m (listM α) → listM α
+  uncons : [Monad m] → listM α → m (Option (α × listM α))
+
+instance : Nonempty (Spec m) := .intro
+  { listM := fun _ => PUnit
+    nil := ⟨⟩
+    cons' := fun _ _ => ⟨⟩
+    squash := fun _ => ⟨⟩
+    uncons := fun _ => pure none }
+
+private unsafe inductive ListMImpl (α : Type u) : Type u
+  | nil : ListMImpl α
+  | cons' : α → ListMImpl α → ListMImpl α
+  | squash : m (ListMImpl α) → ListMImpl α
+
+private unsafe def unconsImpl [Monad m] :
+    ListMImpl m α → m (Option (α × ListMImpl m α))
+  | .nil => pure none
+  | .squash t => t >>= unconsImpl
+  | .cons' x xs => return (x, xs)
+
+@[inline] private unsafe def specImpl : Spec.{u} m where
+  listM := ListMImpl m
+  nil := .nil
+  cons' := .cons'
+  squash := .squash
+  uncons := unconsImpl m
+
+@[implemented_by specImpl]
+private opaque spec : ListM.Spec m
+
+end ListM
+
+/-- A monadic lazy list, controlled by an arbitrary monad. -/
+def ListM (m : Type u → Type u) (α : Type u) : Type u := (ListM.spec m).listM α
+
+namespace ListM
+
+variable {α β : Type u} {m : Type u → Type u}
+
+/-- The empty `ListM`. -/
+@[inline] def nil : ListM m α := (ListM.spec m).nil
+
+/--
+Constructs a `ListM` from head and tail.
+-/
+@[inline] def cons' : α → ListM m α → ListM m α := (ListM.spec m).cons'
```

Comment:
I think this one should be called `cons` and the other one `cons'`, since the monadic one has a different type and also takes an `Option α` head, it's not even a regular list cons.

Reply:
Changed to `cons` and renamed the other to `consOption`. No need for inscrutable primes.